File: src/main/java/textGame/Scene.java
package textGame;

import java.util.Hashtable;
import java.util.Random;

public class Scene extends GameObject implements Searchable, Lootable {
    private static int counter = 0;
    private static Hashtable<Position, Scene> sceneHashtable = new Hashtable<Position, Scene>();
    private int danger;
    private int value;
    private Player player;
    private Inventory inventory = new Inventory();
    private Position position;
    private Npc[] npcs = new Npc[0];

    /// scenes connected to current scene in space 0 -> front, 1 -> right, 2 -> back, 3 -> left
    private Boolean[] openExits = {true, true, true, true};

    public static Scene getByPosition(Position position) {
        return sceneHashtable.get(position);
    }
    public static Scene getByPosition(int x, int y) {
        return getByPosition(new Position(x, y));
    }
    public Position getPosition() {
        return this.position;
    }
    public Player getPlayer() {
        return player;
    }
    public void setPlayer(Player player) {
        this.player = player;
    }
    public void addNpc(Npc npc) {
        Npc[] newNpcs = new Npc[npcs.length + 1];
        for (int i = 0; i < npcs.length; i++) {
            newNpcs[i] = npcs[i];
        }
        newNpcs[newNpcs.length-1] = npc;
        npcs = newNpcs;
        npc.setScene(this);
    }
    public Npc[] getNpcs() {
        return npcs;
    }
    public boolean isExitOpen(int direction) {
        return openExits[direction];
    }
    public GameObject getGameObjectByName(String name) {
        if (name.equals(player.getName()) || name.equals("self")) {
            return player;
        }
        for (Npc npc : npcs) {
            if (npc.getName().equals(name)) {
                return npc;
            }
        }
        Item item = inventory.getItemByName(name);
        if (item != null) {
            return item;
        }
        return null;
    }
    Scene() {
        counter++;
    }
    public static Scene newRandomScene(Position position) {
        Scene scene = FactoryManager.generateRandomScene();
        scene.position = new Position(position);
        Scene.sceneHashtable.put(scene.position, scene);
        return scene;
    }
    public String toString() {
        return this.getName();
    }
    public void generateFromLocationsCSV(int lineNumber) {
        String locationLine = Utility.readCsvLine("src/main/resources/locations.csv", lineNumber);
        String[] locationInformation = locationLine.split(",");

        setName(locationInformation[0]);
        setDescription(locationInformation[1]);
        this.danger = Integer.parseInt(locationInformation[2]);
        this.value = Integer.parseInt(locationInformation[3]);
    }
    public void generateFromRandomLocation() {
        Random rand = new Random();
        int lowerBound = 1;
        int upperBound = Utility.readFileLength("src/main/resources/locations.csv");
        generateFromLocationsCSV(rand.nextInt(lowerBound, upperBound));
    }

    public void removeNpc(Npc npc) {
        if (npcs.length == 0) {
            return;
        }
        if (npcs.length == 1) {
            if (npcs[0] == npc) {
                npcs = new Npc[0];
                return;
            }
        }
        Npc[] newNpcs = new Npc[npcs.length - 1];
        int offset = 0;
        for (int i = 0; i < npcs.length - 1; i++) {
            if (npc == npcs[i] && offset == 0) {
                offset++;
            }
            newNpcs[i] = npcs[i+offset];
        }
        if (offset == 1) {
            newNpcs[newNpcs.length - 1] = npc;
            npcs = newNpcs;
        }
    }

    public Npc getNpc(String name) {
        for (Npc npc : npcs) {
            if (npc.getName().equals(name)) {
                return npc;
            }
        }
        return null;
    }

    @Override
    public Inventory getInventory() {
        return inventory;
    }
}

File: src/main/java/textGame/Scene.kt
package textGame

import java.util.*

class Scene internal constructor() : GameObject(), Searchable, Lootable {
    companion object {
        private var counter = 0
        private val sceneHashtable = Hashtable<Position, Scene>()
        fun getByPosition(position: Position?): Scene? {
            return sceneHashtable[position]
        }

        fun getByPosition(x: Int, y: Int): Scene? {
            return getByPosition(Position(x, y))
        }

        fun newRandomScene(position: Position): Scene {
            val scene = FactoryManager.generateRandomScene()
            scene.position = Position(position)
            sceneHashtable[scene.position] = scene
            return scene
        }
    }
    private val danger = 0
    private val value = 0
    var player: Player? = null
    private val inventory = Inventory(3)
    var position: Position? = null
        private set
    var npcs: Array<Npc> = emptyArray()
        private set

    /** scenes connected to current scene in space 0 -> front, 1 -> right, 2 -> back, 3 -> left */
    private val openExits = arrayOf(true, true, true, true)

    fun addNpc(npc: Npc) {
        npcs += npc
        npc.scene = this
    }

    fun isExitOpen(direction: Int): Boolean {
        return openExits[direction]
    }

    fun getGameObjectByName(name: String): GameObject? {
        if (name == player!!.getName() || name == "self") {
            return player
        }
        for (npc in npcs) {
            if (npc.getName() == name) {
                return npc
            }
        }
        val item = inventory.getItemByName(name)
        if (item != null) {
            return item
        }
        return null
    }

    init {
        counter++
    }

    override fun toString(): String {
        return this.getName()
    }

    fun removeNpc(npc: Npc) {
        npcs = npcs.filter{it != npc}.toTypedArray()
    }

    fun getNpc(name: String): Npc? {
        return npcs.find{it.getName() == name}
    }

    override fun getInventory(): Inventory {
        return inventory
    }
    fun getMapDepiction(): String? {
        return "X"
    }
}